### ISSUE TITLE
Implement a way to add global interceptors

### ISSUE THREAD

**Issue by p-himik** [day8/re-frame#570]
Perhaps something that would be used like
```clj
(re-frame.core/swap-global-interceptors! conj check-db-spec)
```

It will allow having global interceptors that span not only the code that you control (where you can just use your own version of `reg-event-*`) but in any code that you use.
Some good usages for it:
- Checking the DB spec
- Logging
- Adding effects that must be used whenever some DB value changes - no matter where the changes is coming from

Note that is should be discouraged for any library to change the global interceptors set. Only the application itself should do that, while libraries can provide some implementation of the desired interceptors.

**Comment by p-himik** [comment]
The last usage of reacting to a specific DB change made from an arbitrary place (but still via re-frame events and not via a direct `swap!` of the `app-db` atom) is of special interest. It will allow for simpler and more robust code reusability where you need to react to something.

Right now, if I need to emit event `:e` in response to any change of the value under key `:x` in DB, I have to:
1. Create an event handler that changes `:x`, e.g. `:set-x`. Apart from changing `:x`, it will also dispatch `:e`
2. In *every single place* that wants to change `:x`, I now must use `:set-x` instead. I can no longer use `reg-event-db`, I must use `reg-event-fx`. I cannot skip any single event handler that changes `:x` - a maintenance hell, because it also means that I must somehow make sure that such event handlers don't appear in the future.
3. Make sure that I don't use any re-frame library that can change `:x` or can be told to change `:x` but cannot do it via emitting the `:set-x`.

With the proposed change, I will have to:
1. Add a global interceptor that checks if `:x` was changed and emits `:e` if so.

That's it. No hassle whatsoever.

**Comment by p-himik** [comment]
There is a [doodle by Mike Thompson](https://gist.github.com/mike-thompson-day8/7c2d7e770b7bb6ff9627b08c3903a449) that describes an alternative approach.

I think that it doesn't really help us in any way except for removing the point 1 above. With that solution, I will have to:
1. In *every single place* that wants to change `:x`, use `:db-crud` instead of `:db`. It will become much more cumbersome (try to express `merge` in CRUD terms; or some multi-step iterative operation that relies on its previous steps). Every other issue from the point 2 above applies as well.
2. Same as point 3 above. I will not be able to use any library that doesn't use `:db-crud`.

**Comment by mike-thompson-day8** [comment]
So the interesting questions are:
1. Where in the interceptor chain should he global interceptor be added? Immediately next to the handler?  Or instead next to the interceptor whose `:after`  looks after effects? 
2. Need to cater for two or more of these global interceptors being added
3. How about if new global interceptors are added AFTER all event handlers are already registered. Remember, when you register an event handler you create the interceptor chain at that time.  Not later when the event is handled. 
4. Can global interceptors be unregistered?

**Comment by p-himik** [comment]
Awesome!

1. I'd say at the very beginning. This would allow using the `path` and similar interceptors without interfering with the global ones
2. It would be good to have, yes
3. Not sure. Some events can be registered long after the code in `main` is executed - e.g. because of code splitting into multiple modules that are loaded because of user actions. IMO global interceptors should be treated in a special way, without altering the interceptor vector. So, as soon as you register a global interceptor, it should be automatically used both for the already registered events and for the ones that may be registered in the future. Otherwise, there will be ordering issues.
4. I'd say it would be good to have as well

**Comment by superstructor** [comment]
Our proposal is:
1. re-frame provides an API to register a "global interceptor" `reg-global-interceptor`
2. This API call can be called many times to add many global interceptors
3. This API call, internally, conj's each global interceptor onto a global interceptor chain; i.e. adds at the end
4. There will be a new interceptor `run-global-interceptors`
5. This interceptor will be placed into all event interceptor chains at the time of registration, next to the effects interceptor (far left of the interceptor chain)
6. The `:before` of `run-global-interceptors` will run through all currently registered global interceptors and run their `:before`, thus effectively folding them in the interceptor chain. Same for `:after`.
7. There is no ordering guarantee for the running on global interceptors.
8. re-frame provides an API to unregister global interceptors by id

We also considered an alternative API of giving the user full control over the global interceptor chain vector; e.g. `(re-frame.core/swap-global-interceptors! conj check-db-spec)`. We ruled this out as:
- There is no real guarantee of control; e.g. over ordering, what global interceptors exist etc. If you are not the last to call the swap! fn then someone else could reposition your global interceptor in the chain. So its a fake guarantee.
- There are not clear use cases for maintaining full user control over the global interceptor chain.
- Its easier to add it later as an additional API if someone has a requirement than take it away if we release it now.

**Comment by p-himik** [comment]
> There is no ordering guarantee for the running on global interceptors.

Can you please share the reasoning behind not providing such guarantee?
I realize that I cannot possibly expect a newly registered interceptor to be The Last One because there may be some third-party code that also registers them. But I would definitely like to see *my* global interceptors to run in the order I myself call `reg-global-interceptor`.

**Comment by mike-thompson-day8** [comment]
We're seeking to expose the least possible amount of API to get the job done.  Which is why we said "no ordering".  

But the API can obviously be expanded to provide more guarantees if there are good usecases.  Can you share yours? Just so we understand where you are coming from.

**Comment by p-himik** [comment]
From the top of my head - anything (for example, `on-change`) + tracing. It's OK if I can't trace thirdparty code, because I cannot fathom how anyone could guarantee anything here without creating a very convoluted API. But I would prefer to trace at least my own code.

**Comment by superstructor** [comment]
@p-himik in #593 we've got a solution that maintains the order in which `reg-global-interceptor` was called by putting it in a PersistentQueue (i.e. FIFO). I think that meets your suggested requirements re ordering ?

**Comment by p-himik** [comment]
Absolutely, thank you!

**Comment by superstructor** [comment]
Fixed in f817bc9 and released in `1.0.0-rc1`. `1.0.0` will be out shortly.
